backend.native/tests/runtime/concurrent/shared1.kt

```diff
@@ -0,0 +1,508 @@
+    assertEquals(3, global5.get().a)
+
+    global5.dispose()
+    assertFailsWith<NullPointerException> {
+        global5.get().a
+    }
+}
+
+@Test fun testLocal() {
+    val local = DisposableSharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    val worker = Worker.start()
+    val future = worker.execute(TransferMode.SAFE, { local }) { local ->
+        local
+    }
+
+    val value = future.result
+    assertEquals(3, value.get().a)
+    worker.requestTermination().result
+}
+
+@Test fun testLocalDenyAccessOnWorker() {
+    val local = DisposableSharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    val worker = Worker.start()
+    val future = worker.execute(TransferMode.SAFE, { local }) { local ->
+        var result = 0
+        assertFailsWith<IncorrectDereferenceException> {
+            result = local.get().a
+        }
+        result
+    }
+
+    val value = future.result
+    assertEquals(0, value)
+    worker.requestTermination().result
+}
+
+@Test fun testLocalModification() {
+    val semaphore: AtomicInt = AtomicInt(0)
+
+    val local = DisposableSharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    val worker = Worker.start()
+    val future = worker.execute(TransferMode.SAFE, { Pair(local, semaphore) }) { (local, semaphore) ->
+        semaphore.increment()
+        while (semaphore.value < 2) {}
+        local
+    }
+
+    while (semaphore.value < 1) {}
+    local.get().a = 4
+    semaphore.increment()
+
+    val value = future.result
+    assertEquals(4, value.get().a)
+    worker.requestTermination().result
+}
+
+@Test fun testLocalDispose() {
+    val local = DisposableSharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    local.dispose()
+    local.dispose()
+}
+
+@Test fun testLocalAccessAfterDispose() {
+    val local = DisposableSharedRef.create(A(3))
+    assertEquals(3, local.get().a)
+
+    local.dispose()
+    assertFailsWith<NullPointerException> {
+        local.get().a
+    }
+}
+
+fun getWeaksAndAtomicReference(initial: Int): Triple<AtomicReference<DisposableSharedRef<A>?>, WeakReference<DisposableSharedRef<A>>, WeakReference<A>> {
+    val local = DisposableSharedRef.create(A(initial))
+    val localRef: AtomicReference<DisposableSharedRef<A>?> = AtomicReference(local)
+    val localWeak = WeakReference(local)
+    val localValueWeak = WeakReference(local.get())
+
+    assertNotNull(localWeak.get())
+    assertNotNull(localValueWeak.get())
+
+    return Triple(localRef, localWeak, localValueWeak)
+}
+
+@Test fun testCollect() {
+    val (localRef, localWeak, localValueWeak) = getWeaksAndAtomicReference(3)
+
+    localRef.value = null
+    GC.collect()
+
+    assertNull(localWeak.get())
+    assertNull(localValueWeak.get())
+}
+
+@Test fun testDisposeAndCollect() {
+    val (localRef, localWeak, localValueWeak) = getWeaksAndAtomicReference(3)
+
+    localRef.value!!.dispose()
+    GC.collect()
+
+    assertNotNull(localWeak.get())
```

Comment:
This is fragile. Maybe at least create explicit local `val` for `localRef.value!!`?

Reply:
Added a comment that hopefully explains the reasoning. However, variable naming is unfortunate, I'll try to improve it.